### ISSUE TITLE
Is this a memory leak? (very simple WASM module)

### ISSUE THREAD

**Issue by danbugs** [emscripten-core/emscripten#14938]
![image](https://user-images.githubusercontent.com/39843321/130688161-17ffe028-29ca-4856-88ff-990b0b9237b3.png)

This graph shows the JS Heap (from Chrome Dev Tools) across 5 minutes of running the following code:
```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main() {
    char* test = malloc(sizeof(char) *4);
    strcpy(test, "abc");
    printf("%s\n", test);
    free(test);
}
``` 

... as a WASM module.

I've compiled it like: `emcc .\main.c -o index.js`, and my HTML looks like this:
```
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <h1>Test</h1>
    <script async type="text/javascript" src="index.js"></script>
</body>
</html>
```

I don't see any sources for memory leaks but the JS Heap memory usage keeps increasing. Is this expected behavior?

**Comment by sbc100** [comment]
No, I don't see how the wasm module could be leaking any memory.  I will be done running after a new milliseconds once main returns and then does nothing.

What is more WebAssembly code cannot cause leaks like this (at least not in the way most people think) because the heap used by malloc/free is allocated all at once, so even if you have leaks of `malloc`'d blocks they won't show up here since the WebAssembly memory has a fixed size (in general).

**Comment by danbugs** [comment]
Thanks for the quick reply, @sbc100 . Considering what you said, is it normal for memory to increase like this then?

Also, out of curiosity, is there another way you'd recommend I look for memory leaks from mallocs in WASM modules? 🧐

**Comment by sbc100** [comment]
If you want to find memory leaks within your wasm memory we have `-fsantitize=leak` or `-fsantize=address` (the later includes the former plus more).  

See:

https://clang.llvm.org/docs/LeakSanitizer.html

And:

https://emscripten.org/docs/debugging/Sanitizers.html

**Comment by sbc100** [comment]
I'm afraid I don't know why your JS heap might be growing but I don't think it is emscripten related.  There are probably other tools to figure out what is actually causing the growth there.

**Comment by danbugs** [comment]
Thank you for the recommendations on catching memory leaks!

I've just got some quick questions on their use, tho -- Like, if I had a side module, would I have to compile both of them w/ the fsanitize=address flag? Also, if I'm freeing some allocations from the main module in the side module, would it be able to catch that? I'm guessing it wouldn't if I'm linking them through the dynamicLibraries way, huh? Similarly, is it able to catch that I'm not memory leaking things if I'm freeing from the JS side?

**Comment by sbc100** [comment]
Yes, generally you should compiler your entire applications with those flags.   

Sadly it looks like dynamic linking is not yet supported with lsan and asan so you would need to make a static version of you app when running this mode:

```
$ emcc hello.c -fsanitize=address -s SIDE_MODULE
emcc: error: ASan does not support dynamic linking
```

**Comment by danbugs** [comment]
Gotcha -- I'm guessing that's a no from catching mallocs and frees across main and side modules 😅

Still, does it catch mallocs and frees across c/c++ and JS, tho?

Also, when you say static, do you mean just compiling the side module like a main module then?

**Comment by sbc100** [comment]
Yes I meant statically linking your entire app rather than using side or main modules.

I looked into it and it looks like we may be able to enable enable support for `-fsanitize=address` with dynamic linking without too much trouble.

JS and C++ both use the same (native) malloc/free implementation, so yes lsan will catch leaks that occur from either JS or native code.

**Comment by danbugs** [comment]
Gotcha!

Woah — That's awesome! Will it be coming out in an upcoming release then? ;D

Also, out of curiosity, how much trouble would it be supporting EM_JS in side modules too? That's something I've wanted to do before but got the `emcc: error: EM_JS is not supported in side modules` message 😅

**Comment by sbc100** [comment]
I'm not sure how much work adding EM_JS support for side modules would be.  I imagine it would involve using eval and extracting the JS strings from the wasm module.. can you open a bug about that if there is not one already (I would guess there is one).

**Comment by danbugs** [comment]
Hmmm ~ the last mention I'm finding of this is pull request #13649 . Should I create a new issue?

tbh, I'd love to implement the feature myself but I haven't contributed to Emscripten before so I'm not too sure where I'd start looking or even if this is a good first issue 😅

**Comment by danbugs** [comment]
**Update**

I've statically linked my app and compiled w/ `-fsanitize=address`. The result was this:

```
==42==ERROR: AddressSanitizer: dynamic-stack-buffer-overflow on address 0x10968873 at pc 0x0001943e bp 0x10968750 sp 0x1096875c
WRITE of size 1 at 0x10968873 thread T0
    #0 0x1943e in strcpy ../../../../../b/s/w/ir/k/install/emscripten/system/lib/libc/emscripten_asan_strcpy.c:7:15

Address 0x10968873 is located in stack of thread T0
SUMMARY: AddressSanitizer: dynamic-stack-buffer-overflow ../../../../../b/s/w/ir/k/install/emscripten/system/lib/libc/emscripten_asan_strcpy.c:7:15 in strcpy
Shadow bytes around the buggy address:
  0x0212d0b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00       
  0x0212d0c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0212d0d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00       
  0x0212d0e0: 00 00 00 00 00 00 00 00 00 00 00 00 ca ca ca ca       
  0x0212d0f0: 05 cb cb cb cb cb cb cb ca ca ca ca 00 00 00 03       
=>0x0212d100: cb cb cb cb ca ca ca ca 00 00 00 00 00 00[03]cb       
  0x0212d110: cb cb cb cb f1 f1 00 f2 f2 f2 00 00 00 00 f2 f2       
  0x0212d120: f2 f2 00 00 00 00 00 00 00 00 00 00 00 00 00 00       
  0x0212d130: 00 00 f3 f3 f3 f3 00 00 00 00 00 00 00 00 00 00       
  0x0212d140: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00       
  0x0212d150: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00       
Shadow byte legend (one shadow byte represents 8 application bytes):  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==42==ABORTING
program exited (with status: 1), but EXIT_RUNTIME is not set, so halting execution but not exiting the runtime or preventing further async execution (build with EXIT_RUNTIME=1, if you want a true shutdown)
```

Looking at the example from the [Debugging with Sanitizers
](https://emscripten.org/docs/debugging/Sanitizers.html?highlight=sanitizing) section from the Emscripten documentation, I'm not too sure what to make of this. Is the error in a `strcpy` from `emscripten_asan_strcpy.c:7:15`? 😅
